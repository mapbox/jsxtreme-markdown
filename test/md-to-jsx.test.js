'use strict';

const prettier = require('prettier');
const mdToJsx = require('../lib/md-to-jsx');

describe('mdToJsx', () => {
  test('with simple expression', () => {
    const text = `
      # Title
      And a **special**: {{ number }}, {{twice}}.
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('expression inside a link url', () => {
    const text = `
      Go to [the place](/the/place/)

      Go to [the place]({{ thePlaceUrl }})

      Go to [the place](/the/{{ placeUrl }})

      Go to [the place]({{ theUrl }}/place)
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('expression inside a src url', () => {
    const text = `
      ![The image](/the/img.jpg)

      ![The image]({{ theImgUrl }}.jpg)

      ![The image](/the/{{ imgUrl }})

      ![The image]({{ theUrl }}/img.jpg)
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('with nested JSX', () => {
    const text = `
      This is a paragraph {{<span className="foo">with a span inside</span>}}

      {{ <div style={{ margin: 70 }}>
        And here is a div.
      </div> }}
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('with broken-up nested JSX', () => {
    const text = `
      This is a paragraph {{ <span className="foo"> }}with a **markdown** span inside{{ </span> }}

      {{ <div style={{ margin: 70 }}> }}
        And here is a paragraph inside a div.
        [Link](/some/url)
      {{ </div> }}
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('with alternative delimiters', () => {
    const text = `
      This time there's {{ adjective }} new delimiters.
      {{ <div style={{ margin: 70 }}>
        Did this work?
      </div> }}
    `;

    const options = {
      delimiters: ['{{', '}}']
    };

    const jsx = mdToJsx(text, options);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('syntax highlighting clashing with delimiters', () => {
    const text = `
      I'm thinking of a {{color}}

      \`\`\`javascript
      <div style=#{{ color: pink }}>{{ color }}</div>
      \`\`\`
    `;
    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('documentation example', () => {
    const text = `
      # Title
      Here is some **markdown**. So *easy* to write.
      You can interpolate JS expressions like {{ data.number }}
      or {{ dogs.map(d => d.name).join(', ') }}.
      You can also interpolate JSX elements,
      whether {{ <span>inline</span> }} or as a block:
      {{ <div className="fancy-class">
        This is a block.
      </div> }}

      You can even break up JSX interpolation to process more or your text
      as Markdown.

      {{ <div className="fancy-class"> }}
        This is a **Markdown** paragraph inside the div.
        And this is another.
      {{ </div> }}
    `;

    const jsx = mdToJsx(text);
    expect(prettier.format(jsx)).toMatchSnapshot();
  });

  test('error on block-level element within paragraph', () => {
    const text = `
      Text {{ <ul><li>block</li></ul> }}

      More text.
    `;
    expect(() => mdToJsx(text)).toThrow('block-level element');
  });

  test('error on block-level element starting paragraph', () => {
    const text = `
      {{ <ul><li>block</li></ul> }} text.
    `;
    expect(() => mdToJsx(text)).toThrow('block-level element');
  });
});
