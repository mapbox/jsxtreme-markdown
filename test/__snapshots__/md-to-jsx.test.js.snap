// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mdToJsx documentation example 1`] = `
"<div>
  <h1>Title</h1>
  <p>
    {\\" \\"}Here is some <strong>markdown</strong>. So <em>easy</em> to write.
    You can interpolate JS expressions like {data.number}
    or {dogs.map(d => d.name).join(\\", \\")}.
    You can also interpolate JSX elements,
    whether <span>inline</span> or as a block:
  </p>
  <div className=\\"fancy-class\\">
    This is a block.
  </div>
  <p>
    {\\" \\"}You can even break up JSX interpolation to process more or your text
    as Markdown.
  </p>
  <div className=\\"fancy-class\\">
    <p>
      {\\" \\"}This is a <strong>Markdown</strong> paragraph inside the div.
      And this is another.
    </p>
  </div>
</div>;
"
`;

exports[`mdToJsx expression inside a link url 1`] = `
"<div>
  <p>Go to <a href=\\"/the/place/\\">the place</a></p>
  <p>Go to <a href={\`\${thePlaceUrl}\`}>the place</a></p>
  <p>Go to <a href={\`/the/\${placeUrl}\`}>the place</a></p>
  <p>Go to <a href={\`\${theUrl}/place\`}>the place</a></p>
</div>;
"
`;

exports[`mdToJsx expression inside a src url 1`] = `
"<div>
  <p><img src=\\"/the/img.jpg\\" alt=\\"The image\\" /></p>
  <p><img src={\`\${theImgUrl}.jpg\`} alt=\\"The image\\" /></p>
  <p><img src={\`/the/\${imgUrl}\`} alt=\\"The image\\" /></p>
  <p><img src={\`\${theUrl}/img.jpg\`} alt=\\"The image\\" /></p>
</div>;
"
`;

exports[`mdToJsx syntax highlighting clashing with delimiters 1`] = `
"<div>
  <p>I'm thinking of a {color}</p>
  <pre>
    <code className=\\"language-javascript\\">
      &lt;div style={\\"{\\"}{\\"{\\"} color: pink {\\"}\\"}{\\"}\\"}&gt;{color}&lt;/div&gt;{\\"\\\\n\\"}
    </code>
  </pre>
</div>;
"
`;

exports[`mdToJsx with alternative delimiters 1`] = `
"<div>
  <p>
    This time there's {adjective} new delimiters.
  </p>
  <div style={{ margin: 70 }}>
    Did this work?
  </div>
</div>;
"
`;

exports[`mdToJsx with broken-up nested JSX 1`] = `
"<div>
  <p>
    This is a paragraph{\\" \\"}
    <span className=\\"foo\\">with a <strong>markdown</strong> span inside</span>
  </p>
  <div style={{ margin: 70 }}>
    <p>
      {\\" \\"}And here is a paragraph inside a div.
      <a href=\\"/some/url\\">Link</a>
    </p>
  </div>
</div>;
"
`;

exports[`mdToJsx with nested JSX 1`] = `
"<div>
  <p>This is a paragraph <span className=\\"foo\\">with a span inside</span></p>
  <div style={{ margin: 70 }}>
    And here is a div.
  </div>
</div>;
"
`;

exports[`mdToJsx with simple expression 1`] = `
"<div>
  <h1>Title</h1>
  <p>And a <strong>special</strong>: {number}, {twice}.</p>
</div>;
"
`;
